File: July-LeetCoding-Challenge/01-Arranging-Coins/Arranging-Coins.java
/**
 * We can treat it like a problem of Arithmetic progression with a1 = 1, d = 1
 * The sum of this arithmetic sequence is i(i+1)/2, when the sum is first big than we can get that
 * the maxmum staircase shap is i - 1
 */
class Solution {
    public int arrangeCoins(int n) {
        if(n == 0 || n == 1) return n;
        
        int m = 0;
        for(int i = 2; i <= n; i++) {
            if((Math.pow(i, 2) + i) / 2 > n) return (i-1);
        }
        
        return m;
    }
}

File: July-LeetCoding-Challenge/01-Arranging-Coins/Arranging-Coins.cs
public class Solution
{
    public int ArrangeCoins(int n)
    {
        return (int)((Math.Sqrt(8 * (long)n + 1) - 1) / 2);
    }
}

/*
n stars on row x:
x = 1, n = 1
x = 2, n = 1 + 2 = 3
x = 3, n = 1 + 2 + 3 = 6
x = 4, n = 1 + 2 + 3 + 4 = 10

formula: n = (x * (x + 1)) / 2

calculation steps:
2n = x * (x + 1) = x2 + x = x2 + x + 1/4 - 1/4 = (x + 1/2)2 - 1/4
2n + 1/4 = (x + 1/2)2
(8n + 1)/4 = (x + 1/2)2
x + 1/2 = sqrt((8n + 1)/4)
x = sqrt((8n + 1)/4) - 1/2
x = sqrt(8n +1)/2 - 1/2
x = (sqrt(8n +1)-1)/2

*/